Mathlib/Testing/SlimCheck/Functions.lean

```diff
@@ -0,0 +1,540 @@
+  | TotalFunction.withDefault m y, x => (m.dlookup x).getD y
+#align slim_check.total_function.apply SlimCheck.TotalFunction.apply
+
+/-- Implementation of `has_repr (total_function α β)`.
+
+Creates a string for a given `finmap` and output, `x₀ ↦ y₀, .. xₙ ↦ yₙ`
+for each of the entries. The brackets are provided by the calling function.
+-/
+def reprAux [Repr α] [Repr β] (m : List (Σ _ : α, β)) : String :=
+  String.join <|
+    -- Porting note: No `List.qsort`, so convert back and forth to an `Array`.
+    Array.toList <| Array.qsort (lt := fun x y => x < y)
+      (m.map fun x => s!"{(repr <| Sigma.fst x)} ↦ {repr <| Sigma.snd x}, ").toArray
+#align slim_check.total_function.repr_aux SlimCheck.TotalFunction.reprAux
+
+/-- Produce a string for a given `TotalFunction`.
+The output is of the form `[x₀ ↦ f x₀, .. xₙ ↦ f xₙ, _ ↦ y]`.
+-/
+protected def repr [Repr α] [Repr β] : TotalFunction α β → String
+  | TotalFunction.withDefault m y => s!"[{(reprAux m)}_ ↦ {repr y}]"
+#align slim_check.total_function.repr SlimCheck.TotalFunction.repr
+
+instance (α : Type u) (β : Type v) [Repr α] [Repr β] : Repr (TotalFunction α β) where
+  reprPrec f _ := TotalFunction.repr f
+
+/-- Create a `finmap` from a list of pairs. -/
+def List.toFinmap' (xs : List (α × β)) : List (Σ _ : α, β) :=
+  xs.map Prod.toSigma
+#align slim_check.total_function.list.to_finmap' SlimCheck.TotalFunction.List.toFinmap'
+
+section
+
+universe ua ub
+variable [SampleableExt.{_,u} α] [SampleableExt.{_,ub} β]
+
+-- porting note: removed, there is no `sizeof` in the new `Sampleable`
+
+-- /-- Redefine `sizeof` to follow the structure of `sampleable` instances. -/
+-- def Total.sizeof : TotalFunction α β → ℕ
+--   | ⟨m, x⟩ => 1 + @SizeOf.sizeOf _ Sampleable.wf m + SizeOf.sizeOf x
+#noalign slim_check.total_function.total.sizeof
+
+-- instance (priority := 2000) : SizeOf (TotalFunction α β) :=
+--   ⟨Total.sizeof⟩
+#noalign slim_check.total_function.has_sizeof
+
+variable [DecidableEq α]
+
+/-- Shrink a total function by shrinking the lists that represent it. -/
+def shrink {α β} [DecidableEq α] [Shrinkable α] [Shrinkable β] :
+    TotalFunction α β → List (TotalFunction α β)
+  | ⟨m, x⟩ => (Shrinkable.shrink (m, x)).map fun ⟨m', x'⟩ => ⟨List.dedupKeys m', x'⟩
+#align slim_check.total_function.shrink SlimCheck.TotalFunction.shrink
+
+variable [Repr α] [Repr β]
+
+instance Pi.sampleableExt : SampleableExt (α → β) where
+  proxy := TotalFunction α (SampleableExt.proxy β)
+  interp f := SampleableExt.interp ∘ f.apply
+  sample := do
+    let xs : List (_ × _) ← (SampleableExt.sample (α := List (α × β)))
+    let ⟨x⟩ ← (ULiftable.up <|
+      SampleableExt.sample : Gen (ULift.{max u ub} (SampleableExt.proxy β)))
+    pure <| TotalFunction.withDefault (List.toFinmap' <| xs.map <|
+      Prod.map SampleableExt.interp id) x
+  -- note: no way of shrinking the domain without an inverse to `interp`
+  shrink := { shrink := letI : Shrinkable α := {}; TotalFunction.shrink }
```

Comment:
I'm not sure what the right thing to do is here: we only have `Shrinkable (proxy α)` available, but there's no easy way to work with `TotalFunction (SampleableExt.proxy α) (SampleableExt.proxy β)` instead